### ISSUE TITLE
`remove_response` very slow for 24 hours of data - bottleneck in filter-design?

### ISSUE THREAD

**Issue by flixha** [obspy/obspy#2621]
When using the `remove_response` function on a bit longer data segments (e.g., 24 hours of data from one channel), the function becomes quite slow. The following example takes about 90 s for one trace (on Obspy 1.2.1 installed via conda, python 3.7, RHEL 8.3, CPU: Xeon 6242 @2.80GHz). This has earlier been referenced, e.g., in issue #1970. 
(measured with `python -m cProfile -s cumulative test.py`)
```python
from obspy.clients.fdsn import RoutingClient
from obspy import UTCDateTime
client = RoutingClient("eida-routing")
starttime = UTCDateTime("2020-03-01T00:00:00")
endtime = UTCDateTime("2020-03-02T00:00:00")
st = client.get_waveforms(starttime, endtime, network="NS", station="BER",
                          location="00", channel="HHZ")
inv = client.get_stations(starttime=starttime, endtime=endtime,
                          network="NS", station="BER", location="00",
                          channel="HHZ", level="response")
pre_filt = [0.001, 0.005, 45, 50]
st[0].remove_response(inv, pre_filt=pre_filt, output="DISP"))
```
How slow the response-removal is appears to relate to the specific station / response, possibly the number of stages. The following example clocks at just below 30 s:
```python
from obspy.clients.fdsn import RoutingClient
from obspy import UTCDateTime
client = RoutingClient("iris-federator")
starttime = UTCDateTime("2020-03-01T00:00:00")
endtime = UTCDateTime("2020-03-02T00:00:00")
st = client.get_waveforms(starttime, endtime, network="IU", station="KBS",
                          location="10", channel="HHZ")
inv = client.get_stations(starttime=starttime, endtime=endtime,
                          network="IU", station="KBS", location="10",
                          channel="HHZ", level="response")
pre_filt = [0.001, 0.005, 45, 50]
st[0].remove_response(inv, pre_filt=pre_filt, output="DISP")
```

I think there's no computational reason why this function should take so long. cProfile reports that about 90 % of the time are spent in the function `_call_eval_resp_for_frequencies`, and with [line_profiler](https://github.com/pyutils/line_profiler) I saw that 99.7 % of these 90 % are actually spent on this call:
https://github.com/obspy/obspy/blob/09792d405769ca018f3ec089e9b6f52ffeba126e/obspy/core/inventory/response.py#L1585-L1587

I have a harder time profiling the C-library function  `clibevresp._obspy_calc_resp` that is called here, but it appears to me that the C-function increases the computational cost a lot due to its double-loop over all the frequencies at which the response is requested, and over all stages: https://github.com/obspy/obspy/blob/09792d405769ca018f3ec089e9b6f52ffeba126e/obspy/signal/src/evalresp/calc_fctns.c#L61-L70 

I understand that the response needs to be as densely sampled as the seismogram transformed to the frequency domain, so that the two arrays can be multiplied/divided. But if one uses `remove_response` over 1 day of data sampled at 100 Hz, then `calc_resp` needs to loop over 8640000 very tightly spaced frequencies, times the number of stages, which appears unnecesary to me as the response that is returned is usually a rather simple curve. And it costs a lot of computations. 

There's probably several ways of speeding up the call to `calc_resp`, but it would be good to hear some experience / opinions about how to best do it, e.g.: Call `calc_resp` for a limited number of frequencies, and then interpolate it to as many frequencies as the seismogram is sampled at in the frequency domain.

**Comment by flixha** [comment]
I now noticed PR #2592, which moves the Evalresp-calculations to within python (thanks a lot for that @amkearns-usgs!). I tested the examples above with that branch, and found that for IU.KBS.10.HHZ, the response-removal became quicker (20 s instead of about 30 s), but for NS.BER.00.HHZ, it became even slower (112 s instead of 90 s).

Output from cProfile (easier to profile now as all is in python!):
```
         1001370 function calls (974619 primitive calls) in 112.752 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   1404/1    0.016    0.000  112.753  112.753 {built-in method builtins.exec}
        1    0.000    0.000  112.753  112.753 test_response2.py:3(<module>)
        1    0.000    0.000  108.426  108.426 <decorator-gen-28>:1(remove_response)
        1    0.001    0.001  108.426  108.426 trace.py:245(_add_processing_info)
        1    0.081    0.081  108.425  108.425 trace.py:2643(remove_response)
        1    0.001    0.001  100.342  100.342 response.py:997(get_response_for_window_size)
        1    0.221    0.221  100.311  100.311 response.py:1039(get_response)
        6    0.472    0.079   99.485   16.581 response.py:771(get_response)
        6    1.037    0.173   99.013   16.502 filter_design.py:275(freqz)
       12   97.975    8.165   97.975    8.165 polynomial.py:643(polyval)
  487/348    0.121    0.000    7.893    0.023 {built-in method numpy.core._multiarray_umath.implement_array_function}
        2    0.094    0.047    7.340    3.670 _pocketfft.py:51(_raw_fft)
.........
```

So I think the issue remains valid that I'm wondering whether we need to compute the response at so many densely spaced frequencies. The slow function call in PR #2592 is now:
https://github.com/amkearns-usgs/obspy/blob/f458b372cbcbb1d2c663d138ff6f598a513ebdf6/obspy/core/inventory/response.py#L785

**Comment by aringler-usgs** [comment]
@flixha #2592 was written to replicate the evalresp implementation as closely as possible.  So it makes sense that it will not be as fast as a C implementation.  From my experience, there are a lot of "odd" things in evalresp that don't make sense to me.  

Some of the "odd" stuff might be related to getting the FIR filters and the high-frequency pieces correct.

I would also be interested in hearing other's perspectives on this.  When @amkearns-usgs implemented it he largely just tried to match as much of evalresp as possible, even if evalresp didn't make sense.

**Comment by amkearns-usgs** [comment]
Yes. While the replaced code relies as much as possible on numpy and scipy for numeric operations (which have the advantage of backends that are mostly C-coded and thus still fairly efficient), there's probably some additional overhead about the way the python and C interacts there -- the evalresp calls are done in basically a single step in the old code, so converting to and from python datatypes is done only once where chained calls to numpy/scipy operations leads to having to do those operations more frequently.

I'm not really a python guru, however (and for that matter prefer to work in C as little as possible, though I have been looking into python-C interaction for the sake of a different project, so it's very fresh in my mind); take this with some grains of salt.

**Comment by krischer** [comment]
@flixha You are correct in that it is a ridiculous overkill to compute the response at every single frequency. Evalresp indeed includes some functionality to compute it on less samples and then interpolate with splines to all frequencies. We never wrapped this but probably should have.

This is a lot simpler to implement in #2592 and thus I would suggest to do it there. I would assume that computing the actual response on maybe 1000 samples is enough (one would have to test of course). I assume that then the FFT will become the limiting factor but there is only so much one can do there although some kind of sliding window approach might also help here

**Comment by flixha** [comment]
Thanks for your replies. Interesting to know that evalresp actually included functionality for interpolation, but it's true that everything (profiling, debugging, etc.) becomes easier with the python-implementation in #2592. I'm not sure whether I should best create a new PR in @amkearns-usgs 's fork; or wait with a PR until #2592 is merged into master?

 Anyways, the code to change things is very short, e.g., in https://github.com/amkearns-usgs/obspy/blob/f458b372cbcbb1d2c663d138ff6f598a513ebdf6/obspy/core/inventory/response.py#L785-L787:

```python
        # Compute response for a limited number of frequencies - 50000 appears
        # suitable for good-enough precision vs. speed
        if len(frequencies) > 50000:
            resp_frequencies = np.linspace(frequencies[0], frequencies[-1],
                                           50000, dtype=np.float64)
            resp = scipy.signal.freqz(b=coefficients, a=[1.],
                                      worN=resp_frequencies)[1]
            amp = np.abs(resp) * self.stage_gain + 0j
            amp = amp.real
            amp = scipy.interpolate.InterpolatedUnivariateSpline(
                    resp_frequencies, amp, k=3)(frequencies)
        else:
            resp = scipy.signal.freqz(b=coefficients, a=[1.], worN=frequencies)[1]
            # Here we zero the phase (FIR) and return the amplitude
            amp = np.abs(resp) * self.stage_gain + 0j
            amp = amp.real
```

Some tests for the number of frequencies for which respeval is called; vs. the maximum difference in the velocity-amplitude of a test trace (for accurate computation vs. spline interpolation); vs. the quickest time I measured (includes reads and imports):

n(freqs) |  max diff in m/s | time
....1000  |..... ~10^-14 ......|  7.76 s
..10000  |.....  ~10^-17 ......| 7.79 s
100000  |..... ~10^-19 ......| 8.40 s
previously, without interpolation:
8640000|----------------------| 120 s

So at these numbers of iterations, the time isn't too different. The test above is for a high-broadband trace. If there were any measurement that span a lot wider bandwith than standard high-broadband, then maybe a user would wish for higher precision. I can't think of a case like that right now, so I'm just wondering whether we should keep an option to  explicitly compute all frequencies, or whether that would be unnecessary clutter?

**Comment by megies** [comment]
> then maybe a user would wish for higher precision. I can't think of a case like that right now, so I'm just wondering whether we should keep an option to explicitly compute all frequencies, or whether that would be unnecessary clutter?

keeping such an option is likely a "why not", should only be a few lines of docstring documentation, and a single line change..

```python
        if len(frequencies) > 50000 and fast:
```

**Comment by d-chambers** [comment]
Hey @flixha, can this be closed?

**Comment by flixha** [comment]
> Hey @flixha, can this be closed?

Yes, this is addressed in the big PRs for switching evalresp to pure python that are linked above.
